### ISSUE TITLE
QT problems thread

### ISSUE THREAD

**Issue by zeromus** [TASVideos/fceux#332]
A brutal lot of feedback, but if my criteria can be met (and I can confirm I can actually dev on the qt frontend without losing my mind and think other devs should be able to also) then we can get it to the point of actually deprecating the win32 frontend apart from tastudio.

- [x] Hex editor is missing .tbl support.
- [x] Hex editor crashes on "view rom" when no rom is open
- [x] Hex editor crashes on "view ppu" when no rom is open
- [x] Default size of hex editor is not good
- [ ] Is this mesen-styled? Can we have some kind of a Color > Mesen Style or Color > FCEUX Style option to change it in one fell swopp?
- [x] Debugger resizes visibly when starting up
- [x] "Use native OS File Dialog" doesn't work
- [x] There's no recent roms list
- [x] Trying to save a state with shift+F1 makes it crash
- [x] Including the redist is a cool idea, but can we consider making it a link to the redist instead?
- [x] The default window size was 512x494 for me. This is weird. In general sizing this stuff is a friggin nightmare
- [x] On windows, the ".fceux" directory should not be used by default. That's just not how most windows fceux users want it to work. I suggest instead default it to a subdirectory ".user" alongside the exe. You may be able to find some best practice in "portable apps" universe to find a way to control it to be "portable" or not. Could be a first-run setting (user files in user directory or portable mode) and then if it's "portable" mode, the local .user stuff is written and that marks it as portable; otherwise write a FILE called .user which is just a dummy marker that means the .user is elsewhere?
- [x] Not that the win32 frontend is fully consistent in this respect, but some dialogs are missing close and/or apply buttons. This is really disorienting to windows users. Sometimes the top-right X button saves and closes, while other times it doesn't (gamepad menu for instance warns you that it won't be saving the settings). If you don't want to do OK and cancel everywhere then we need close everywhere (and top-right X needs to try to work consistently)

**Comment by mjbudd77** [comment]
The Qt/SDL GUI was crashing when F1 button was hit. I have fixed this. However, the F1 hot key is mapped to bringing up the cheat window. Upon comparing the hot keys between the win32 and Qt GUIs, I see a lot of differences.

**Comment by mjbudd77** [comment]
For the windows Qt gui, the config file will be created along side the executable (like Win32 version does). This behavior can be overridden if the user sets an FCEUX_HOME environment variable to point to an alternate storage directory.

**Comment by eugene-s-nesdev** [comment]
"Recent ROMs" and "Reset" issue, step by step, QT-version:
1. Open emulator, "File" -> "Recent Roms" -> ROM1.NES
2. "File" -> "Recent Roms" -> ROM2.NES
3. Press "Emulation" -> "Reset". You'll see ROM1 instead of ROM2.
BTW, "Power" and "Soft Reset" buttons works correctly, and will play ROM2, as needed.

**Comment by mjbudd77** [comment]
Recent ROM hard reset issue should now be fixed in commit 8e4282f.

**Comment by zeromus** [comment]
The redist doesn't affect whether it runs on any computer. It is just a waste of space. Some computers will need to _run_ it before qtfceux will work. Those people will still need us to tell them to do that. Via a web page, at which time we can provide the link to the redist.

**Comment by mjbudd77** [comment]
After reading about windeployqt more, I found that it has a convenient command line option --no-compiler-runtime that will omit the redist from the collected binary files. I added this in commit c962e2b. Still TODO add redist information to the web page.

**Comment by mjbudd77** [comment]
All windows in Qt GUI should now have either close buttons at bottom of window or File -> Close menu bar options. Please let me know if I have missed one.

**Comment by SonnyJim** [comment]
Whenever I try to run the latest git on my Sparkylinux laptop (Debian/LXQt) it just gives me black on black menu options.  Rolling back the code to the 2.3.0 release (5f1263a) gives me the menus back, so not totally sure what's different.

**Comment by mjbudd77** [comment]
Did you try deleting your saved config file? $HOME/.fceux/fceux.cfg. The development version has new styling options that may have bogus values.

**Comment by SonnyJim** [comment]
Just rm -rf'ed ~/.fceux and tried again, still black on black :(

![image](https://user-images.githubusercontent.com/123852/115145197-b4ddbf00-a093-11eb-917d-488377697232.png)

**Comment by mjbudd77** [comment]
Try this, go to options -> UI config in main menu. This will open a new window. Choose fusion base style

**Comment by SonnyJim** [comment]
Yeah I managed to navigate there blindly, it did segfault when I changed it but fortunately the change stuck and seems to be working now, thanks.

**Comment by mjbudd77** [comment]
Just curious, Which style was it using before? I think I’m going to force the default style to be fusion for Linux systems. Right now it tries to get the default style from desktop environment which isn’t always the best choice

**Comment by SonnyJim** [comment]
bb1black (or similar).  I'm pretty sure it's my weird choice of desktop environment that's causing the issue.

**Comment by SonnyJim** [comment]
Found another small bug, not sure if it affects the Windows version.  If I change the trace log size whilst the trace log is active and emulation is paused it causes a seg fault.

**Comment by mjbudd77** [comment]
I just pushed a commit that fixes the trace log crash issue.

**Comment by SonnyJim** [comment]
Yep working fine now.  Thanks.

I am suffering from a very strange issue where some of my symbols are 'disappearing' from $title.nes.ram.nl.  From what I can gather, it's related to when I've manually edited a ROM address in the hex editor window.

**Comment by mjbudd77** [comment]
That one is strange. Since this bug is going to be more involved, I suggest that a new issue be opened and more information be provided.

**Comment by SonnyJim** [comment]
Not a problem.  A quick question to save me wading through the code and to help me diagnose the problem quicker:

When does FCEUX write the label to disk?  Is it as soon as I create the label (ie right click address, Add Symbolic Debug marker, enter text, OK), or when I close the ROM/emulator?

**Comment by mjbudd77** [comment]
I believe it writes to disk when the ROM is closed. CloseGame function in fceuWrapper.cpp calls debugSymbolTable.save().

**Comment by SonnyJim** [comment]
Ah that might explain it then.  I've had a few segfaults so it's likely that it's just not been saved.  Would it make more sense to call debugSymbolTable.save() on each change?  I can't see it being a huge performance penalty.

**Comment by mjbudd77** [comment]
I do not see that as a big performance penalty either. I will look into save on every change. I am concerned about these segmentation faults you are experiencing.  If you know how to reproduce them, I would love to know how to trigger the faults and see some process image dumps/stack traces of the faults. I want to make the program as stable as possible

**Comment by SonnyJim** [comment]
Here's one:

Open ROM, Open debugger window
press 'Step Into' to pause the emulation 
Select 'Close ROM' from the file menu.
Press 'Run' in the debugger window.

At a guess, Close ROM should also close the debugger window even if the emulation is paused.

**Comment by SonnyJim** [comment]
There's also a bug with the hex editor window which is kinda hard to explain:

Open ROM, debugger window
Scroll to a 'high' address, in my example I'm using Bomber King and $B344.
Right click that address, select 'Open in Hex editor'
This takes the Hex Editor window to 0x1B354, now if I scroll up to 0x0000 it shows the file header ("NES...) rather than the RAM area.  To get it 'back to normal' I have to select a 'low' address in the debugger and select 'Open in Hex Editor', this puts it back to showing the NES RAM rather than the file contents.

I'm pretty sure this is expected behaviour, but it would be nice if the Hex Editor window title could reflect whether I'm looking at NES RAM or file contents.

**Comment by SonnyJim** [comment]
Using 'Unfreeze all' in the Hex Editor whilst the emulation is paused via the debugger:

Open ROM, Debugger and Hex Editor window
Pause the emulation via Debugger 'Step Into'
Pick a memory location in the hex editor window, right click and freeze the address
Right click again, select "Unfreeze all'.
It should segfault

**Comment by mjbudd77** [comment]
> Here's one:
> 
> Open ROM, Open debugger window
> press 'Step Into' to pause the emulation
> Select 'Close ROM' from the file menu.
> Press 'Run' in the debugger window.
> 
> This segfaults everytime for me. At a guess, Close ROM should also close the debugger window even if the emulation is paused.

This should be fixed in commit b97bf287b2b4e4e4c1978df16fc4cedcea6e3856

**Comment by mjbudd77** [comment]
> Using 'Unfreeze all' in the Hex Editor whilst the emulation is paused via the debugger:
> 
> Open ROM, Debugger and Hex Editor window
> Pause the emulation via Debugger 'Step Into'
> Pick a memory location in the hex editor window, right click and freeze the address
> Right click again, select "Unfreeze all'.
> It should segfault

This should be fixed in commit 974d2aefba85ea7b787984e83142232e20512565

**Comment by mjbudd77** [comment]
> There's also a bug with the hex editor window which is kinda hard to explain:
> 
> Open ROM, debugger window
> Scroll to a 'high' address, in my example I'm using Bomber King and $B344.
> Right click that address, select 'Open in Hex editor'
> This takes the Hex Editor window to 0x1B354, now if I scroll up to 0x0000 it shows the file header ("NES...) rather than the RAM area. To get it 'back to normal' I have to select a 'low' address in the debugger and select 'Open in Hex Editor', this puts it back to showing the NES RAM rather than the file contents.
> 
> I'm pretty sure this is expected behaviour though the color highlighting seems to go a bit weird.

I believe this is expected behavior. It is a bad idea to attempt to edit ROM memory that is mapped to the CPU address space. For this reason, the hex editor opens up the ROM view (where it is safer to edit) instead of the CPU view.

**Comment by mjbudd77** [comment]
debugSymbolTable.save() is now called after any change to the debug symbol table. Commit ac340e0a4c7bbc8b3b111e58702b2cdae65a6a77

**Comment by SonnyJim** [comment]
Just tested the latest commit, no more segfaults, thank you ;)

**Comment by SonnyJim** [comment]
A small feature request please.  At the moment the options in the Trace Logger aren't saved, meaning I have to turn on Symbolic trace each time.  Would it be possible for the trace logger options be saved as well?

**Comment by mjbudd77** [comment]
Yes, I will make all the trace logger settings be saved parameters.

**Comment by eugene-s-nesdev** [comment]
Tested on EEEPC which lacks OpenGL 2.0 (version 1.4 only supported):
https://youtu.be/j2cXrB8T3js

bug1: SDL video works only after restart video driver on-the-fly (SDL -> OpenGL -> SDL again)
bug2: Aspect ratio cancel to square when changing video deriver. Need to re-apply 4:3 settings
bug3: Turbo buttons default speed don't equal to win32-version, skipping appears
bug4: emulator creates ROMNAME.nes.ram and ROMNAME.cdl files in your ROMs folder. Win32-version doesn't.

**Comment by mjbudd77** [comment]
I have pushed some changes for these bugs. Please recheck 1,2,6, and 7. Also, I looked at the OpenGL version differences and added some checks to use older texture rendering methods if certain extensions are not installed. Please check if OpenGL now works for 1.4.

**Comment by eugene-s-nesdev** [comment]
Bugs 1, 2, 7 are fixed.
Bug 6 is still present. I will more test on another machines.
I will test bug 3 today.

**Comment by mjbudd77** [comment]
@eugene-s-nesdev can you verify that the OpenGL is now functional for machine #1 that has the older OpenGL version?

**Comment by eugene-s-nesdev** [comment]
No, it does't work. GMA3150 is very old chip and have bad opengl support under win7.
I used this netbook a long time ago under lubuntu 14.04, and retroarch-ogl works here (linux have better GMA3150 support). But now, under win7, i cannot find any opengl software that works here, except for retroarch when forcing "GL1" video driver.

**Comment by eugene-s-nesdev** [comment]
Re-check bug 20 please https://github.com/TASVideos/fceux/commit/4ab45b82d833055f33cc2bd4bc50ca9b1f7661bd
I doubt that win32-version have correct behavior here: https://imgur.com/a/Fas3vOd 

What's wrong here? https://github.com/TASVideos/fceux/issues/357

**Comment by mjbudd77** [comment]
I don’t know much about how the internals of the sound work, should I just make all the sliders usable regardless of whether they do anything?

**Comment by eugene-s-nesdev** [comment]
> should I just make all the sliders usable regardless of whether they do anything? 

Yep, i guess.

> Also, who actually uses the low quality sound option?

LQ-option need for low-powered devices, like intel atom, RPI2/RPI3 etc, it give huge performance boost.

**Comment by eugene-s-nesdev** [comment]
> For Number 9, added overclocking option in commit b97b652. Will need some help testing, I'm not sure what to look for other than I could see CPU usage go up slightly when enabling.

Overclock works correct as expected. You can use overclock rom by tepples, or 240pee.nes for test it.
Nice feature is "show lag counter" and "reset lag counter" for visual lag control.

**Comment by mjbudd77** [comment]
@eugene-s-nesdev   I have not been able to reproduce number 25 on my only windows computer, but I did find a known issue on the Qt5 website with OpenGL based windows. So I implemented the recommended changes to fix the issue referenced at this link:  https://doc.qt.io/qt-5/windows-issues.html#fullscreen-opengl-based-windows
Changes are in commit d9f10a96. Try it out and let me know if it helps.

**Comment by eugene-s-nesdev** [comment]
> Changes are in commit d9f10a9. Try it out and let me know if it helps.

Thank you. Bug 25 partially fixed. OpenGL drop-down menus is visible and working.
Still have problems with SDL menu in windowed mode. My old videos shows this issue.
but here is new, if you want it:
https://youtu.be/A54zvhtKkX4 (win7)
https://youtu.be/rJ8-BFOgA1E (win10)

**Comment by mjbudd77** [comment]
Fixed number 26 in commit b1b9b50.

**Comment by pinobatch** [comment]
On Xubuntu 21.04, if I map single quote (`'`) to player 1's Select Button, pressing single quote on my keyboard doesn't press Select in-game. In addition, Right Shift shows up as Left Shift, though that mapping actually works.

**Comment by mjbudd77** [comment]
@pinobatch, you most likely have a hot key conflict. I had the same issue until I went into the hot key config window and cleared the SDL.HotKeys.MuteCapture hot key. I don't know what that hot key actually does but you probably don't need it. Regarding the shift key, Qt does not distinguish between left and right shift. It only tells me the shift modifier is active. I do not recommend using any modifier keys for mapping of gamepad buttons. This includes shift, alt, ctrl, fn, windows, etc.

**Comment by eugene-s-nesdev** [comment]
Here are AIDA64 OpenGL reports for all my machines: 
[OpenGL.zip](https://github.com/TASVideos/fceux/files/6625385/OpenGL.zip)
Maybe it will be helpful for you.

buglist link:
https://github.com/TASVideos/fceux/issues/332#issuecomment-843888724

**Comment by mjbudd77** [comment]
I think I've finally fixed number 25 in commit 2861067.

For number 4, I have code in place to favor selecting gamepads over keyboards if the saved input device is unavailable. How would you prefer for this to work?

**Comment by eugene-s-nesdev** [comment]
Thank you, i will test bug25 for now.
Regarding bug4, i don't know yet how to make correct behavior, but this is https://youtu.be/QF7WACI3oD4 looks like a bug

**Comment by ClusterM** [comment]
The GamePad config dialog is very confusing. It is not obvious at all that a user needs to save the profile to save the current settings. It more confusing when a user sees this message:

![image](https://user-images.githubusercontent.com/4236181/121751630-393e4280-cb17-11eb-9bf4-b1b394e66ec5.png)

But his new settings actually work until the emulator restarted.

Also, the Qt version lacks the feature to bind multiple keys/buttons at the same time without the requirement to switch between profiles. It's very annoying.

**Comment by ClusterM** [comment]
Also, the Hotkey Configuration dialog shows the "Set Hot Key" dialog twice for some reason. Only the last keypress is saved. And I still can't bind gamepad buttons using it (issue #312).

**Comment by mjbudd77** [comment]
I will look into why the hot key dialog shows twice. The Qt GUI game pad to hot key function does not work this way. Try setting a keyboard sequence for a hot key and then go to advanced settings in the game pad config window to bind the game pad button to the key sequence you want to be triggered

**Comment by mjbudd77** [comment]
Number 28 is fixed in commit e79f015.

**Comment by mjbudd77** [comment]
@ClusterM The double hot key set window activation should now be fixed in commit da36b4a.

**Comment by ClusterM** [comment]
> The Qt GUI game pad to hot key function does not work this way. Try setting a keyboard sequence for a hot key and then go to advanced settings in the game pad config window to bind the game pad button to the key sequence you want to be triggered

Wha... it's a kludge. Why can't you just use the same key/input dialog for hotkeys like in the Windows version?

**Comment by mjbudd77** [comment]
It’s not a kludge, it’s actually an elegant and very flexible way to tap into any GUI shortcut. Read about how QAction and QShortcut works. By generating pseudo keyboard inputs I can trigger any shortcut in the GUI. I noticed that the win32 lacks flexibility because it maps generic joystick buttons to the hot key function. Every joystick is different and that is why I put the key bindings in the game pad config. Button zero on one joystick may not be the same button on another. So the plan was to save advanced key bindings with the game pad profile which is not fully done. That functionality is still in development but I didn’t any feedback on it in the other issue thread so I didn’t think anyone cared. I have some ideas to make setting the game pad hot key bindings more intuitive and I am open to suggestions. But I’m convinced that this architecture is the best since it is most flexible and does not duplicate code.

**Comment by ClusterM** [comment]
>Button zero on one joystick may not be the same button on another.

That's why need to store the button number with controller GUID for every binding. There is no feedback on the other thread because there are not so many users of the Qt version yet. I think this will change soon.

**Comment by mjbudd77** [comment]
@ClusterM i have made some changes to the game pad hot key bindings interface. I believe it will be more inline with your expectations. Still have to bind the button to the hot key via the game pad config window, but now I have a hot key selection dialog that the user can just click the function they want to bind to. I don’t have the bindings being saved to a file yet, still figuring out how I want to do that. Also, I will like to have more discussion about how you are using basic game pad setup and how you are managing your various controller profiles. The changes I have made to the game pad config where better align with the way SDL handles things. Maybe there is some common ground we can find if I can better understand your frustrations.

**Comment by mjbudd77** [comment]
I would like to start this discussion by explaining how the setup currently is designed to function and why it was done this way. 
1. All gamepads have a unique GUID associated with them. (As you already know)
2. The SDL2 library contains an internal database of gamepad button mappings for each GUID.
3. For each gamepad GUID, I create a folder where various button mapping profiles can be saved and loaded. A default profile is automatically created if one does not exist for each GUID based the SDL2 database. Most of the time, this makes the default profile usable without having to map a single button manually. Default profiles can be overwritten if desired.
4. The format of the saved button mapping config files a superset of the standard SDL2 database record text format. This ensures compatibility with SDL2 database.
5. Every profile is unique to a GUID, so I can make a profile called “Matt” for each gamepad type. At startup, the system will try to match each player port to the last known GUID and profile name. It first tries to match both GUID and profile. If the GUID cannot be matched, meaning gamepad is not plugged in, then it will try to match the profile name to the first detected gamepad GUID that is not already in use. If this fails, then the first detected gamepad is used with default settings. If no gamepad is found, then keyboard is used first trying to match keyboard profile and then default keyboard profile.

**Comment by ClusterM** [comment]
> I can better understand your frustrations

I'm not frustrated at all :) I just think that it's not user-friendly. The only frustrating thing is a lack of ability to bind multiple keys/buttons at once. It's not a problem when you just playing a game but fceux is a powerful tool for debugging, romhacking, and TAS. This is a fairly common situation when you need to constantly switch between keyboard controls and gamepad controls while using the debugger.

**Comment by eugene-s-nesdev** [comment]
mjbudd77: Thank you very much for color picker / eyedropper tool.
My request is ability to cancel last picked color on right-click mouse.
For now, you can't cancel last operation quickly, need to reset palette to default.

![picker](https://user-images.githubusercontent.com/13199529/121890308-df11cd00-cd22-11eb-8b08-39485cab5072.png)

**Comment by mjbudd77** [comment]
@eugene-s-nesdev I added an Edit -> Undo/Redo operations to the menu of the palette color editor. You can now undo/redo quickly using the keyboard shortcuts shown in the menu text (Ctrl+Z  and Ctrl+Shift+Z).

**Comment by mjbudd77** [comment]
@eugene-s-nesdev Please recheck number 4. On commit 167bda0d37269f21a5c02988184f4e069591be3b

**Comment by eugene-s-nesdev** [comment]
Thank you, it's fixed

**Comment by mjbudd77** [comment]
I added PAL-filter adjusting sliders for number 28. Also added code to save the config settings for number 30.

**Comment by mjbudd77** [comment]
@ClusterM after looking more at the Qt gamepad code in more detail. I think I’m going to postpone changing the code as it will greatly delay making a new release in the near future. Just too many big changes need to be made. I want to get the 2.4.0 release out soon. I have a plan for  another release late this year or early next year that will include porting over the TAS editor. I will lump this change in with that release.

**Comment by eugene-s-nesdev** [comment]
> I added PAL-filter adjusting sliders for number 28

PAL filter checkbox reset to ON every time you restart emulator, unlike NTSC filter and other checkboxes, which remember their last values. It happens every time, even when "none" scaler selected.

**Comment by eugene-s-nesdev** [comment]
> I added PAL-filter adjusting sliders for number 28

PAL filter checkbox reset to ON every time you restart emulator, unlike NTSC filter and other checkboxes, which remember their last values. It happens every time, even when "none" scaler selected.
BTW, win32-build doesn't have this checkbox, maybe this is a reason of this bug

Regarding old bug8: why palette config window have so much unused space?
Maybe better to make something like this?:
![palette](https://user-images.githubusercontent.com/13199529/122479407-b5330180-cfd3-11eb-8328-09246bb02a30.png)

**Comment by mjbudd77** [comment]
@eugene-s-nesdev  I condensed the Qt palette config window layout. Let me know if it is any better.

**Comment by eugene-s-nesdev** [comment]
Yes, it's better. I've just observed that sharpness slider don't have (and never had) percent indicator in QT-version.

**Comment by eugene-s-nesdev** [comment]
Just a small cosmetic fix needed: https://youtu.be/WMjUmhWLEA4

**Comment by mjbudd77** [comment]
@eugene-s-nesdev I added some minimum widths to the PAL sliders and labels which should help with the issue of things shifting with the text size changes. Looks good on a 1080p resolution, you will have to let me know how it looks on the smaller res display.

**Comment by eugene-s-nesdev** [comment]
Thank you all gui elements looks very nice.
Regarding bug 24. I agree, this is strange thing. 
Moreover, i just got intelNUC for testing:
```
intel core i3-5005U
Intel HD Graphics 5500
graphic driver 20.19.15.5171 (dev 04.11.2020), whql
```
and got absolutely same bug24 as intel HD Graphics 4600 (20.19.15.4531) WHQL had
If none helps, you can just switch all windows builds to SDL-driver by default

**Comment by mjbudd77** [comment]
For number 31, Qt does not distinguish between left and right shift. So I cannot know which is used. This is true for all modifier keys. Don't map game pad buttons to shift, ctrl, alt,meta, etc.

**Comment by mjbudd77** [comment]
@eugene-s-nesdev I'm not sure that its all intel hardware that has a problem. I managed to do a test on a Dell Latitude 7280 Windows 10 OS.
```
Intel Core i5-6300 CPU @ 2.40 GHz
Intel HD Graphics 520
Driver version 27.20.100.9171
OpenGL version 4.6.0
```
All appears to working properly with the video. Tested all filters at various scale factors.

**Comment by mjbudd77** [comment]
I would like to thank you all for your help in testing for the 2.4.0 release of the Qt GUI. I have decided to go ahead with releasing the software now. I will continue to improve the GUI moving forward and have a plan in place to eventually port over the TAS editor. I'm closing this thread as it has gotten a little out of control and getting hard to keep track of. If you have specific bugs that you would like me to look into, please open new issues (1 issue per bug/feature request please  :)   )
